### PR TITLE
Update Distributed components and integrate it with FT pipelines and Model Eval fixes

### PR DESCRIPTION
**Pull request by ankushbhatia2**

### CODE REVIEW

File: assets/training/finetune_acft_hf_nlp/components/pipeline_components/chat_completion/spec.yaml [hunk 1]
```diff
@@ -618,7 +618,7 @@ jobs:
       # mlflow_model_folder: '${{parent.outputs.mlflow_model_folder}}'
   chat_completion_model_converter:
     type: command
-    component: azureml:chat_completion_model_converter:0.0.52
+    component: azureml://registries/azureml/components/chat_completion_model_converter/versions/0.0.53
```
Comment (skasturi) [inline assets/training/finetune_acft_hf_nlp/components/pipeline_components/chat_completion/spec.yaml:621]:
```suggestion
    component: azureml:chat_completion_model_converter:0.0.53
```

File: assets/training/finetune_acft_hf_nlp/components/pipeline_components/chat_completion/spec.yaml [hunk 2]
```diff
@@ -627,34 +627,30 @@ jobs:
...
-  #     prediction: '${{parent.jobs.model_prediction.outputs.predictions}}'
-  #     prediction_column_name: predictions
-  #     prediction_probabilities: '${{parent.jobs.model_prediction.outputs.prediction_probabilities}}'
-  #     evaluation_config: '${{parent.inputs.evaluation_config}}'
-  #     evaluation_config_params: '${{parent.inputs.evaluation_config_params}}'
-  #   outputs:
-  #     evaluation_result: '${{parent.outputs.evaluation_result}}'
+  model_prediction:
+    type: command
+    component: azureml:model_prediction_with_container:0.0.2
```
Comment (skasturi) [inline assets/training/finetune_acft_hf_nlp/components/pipeline_components/chat_completion/spec.yaml:632]: Could we rename `model_prediction_with_container`?

File: assets/training/model_evaluation/components/distributed_model_prediction/spec.yaml [hunk 3]
```diff
@@ -1,6 +1,6 @@
 $schema: https://azuremlschemas.azureedge.net/latest/pipelineComponent.schema.json
 name: model_prediction_with_container 
-version: 0.0.1
+version: 7.25.2024.16
```
Comment (skasturi) [inline assets/training/model_evaluation/components/distributed_model_prediction/spec.yaml:3]: Are we introducing a new version scheme?

Comment (ankushbhatia2) [reply assets/training/model_evaluation/components/distributed_model_prediction/spec.yaml]: Yes, it was accidentally added. It was the version I was using for debugging. I have fixed this

File: assets/training/finetune_acft_hf_nlp/components/pipeline_components/chat_completion/spec.yaml (hunk 2 above)
Comment (ankushbhatia2) [reply assets/training/finetune_acft_hf_nlp/components/pipeline_components/chat_completion/spec.yaml:632]: The component version 0.0.1 is already there in prod. This is an upgrade for that component

File: assets/training/finetune_acft_hf_nlp/components/pipeline_components/chat_completion/spec.yaml (hunk 1 above)
Comment (ankushbhatia2) [reply assets/training/finetune_acft_hf_nlp/components/pipeline_components/chat_completion/spec.yaml]: Fixed in the latest commit.

File: assets/training/model_evaluation/src/evaluators/evaluators.py [hunk 4]
```diff
@@ -617,16 +617,32 @@ def evaluate(self, y_test, y_pred, **kwargs):
         """
         #  dataframe with 2 columns predictions and predictions appended to the conversation
         if len(y_pred.columns) > 1:
```
Comment (skasturi) [inline assets/training/model_evaluation/src/evaluators/evaluators.py:619]: Are we validating y_pred to be of the correct type?

File: assets/training/model_evaluation/src/evaluators/evaluators.py [hunk 5]
```diff
@@ -617,16 +617,32 @@ def evaluate(self, y_test, y_pred, **kwargs):
         """
         #  dataframe with 2 columns predictions and predictions appended to the conversation
```
Comment (skasturi) [inline assets/training/model_evaluation/src/evaluators/evaluators.py:618]: (nit) Please add method description

File: assets/training/model_evaluation/src/evaluators/evaluators.py [hunk 6]
```diff
@@ -617,16 +617,32 @@ def evaluate(self, y_test, y_pred, **kwargs):
...
         if len(y_pred.columns) > 1:
-            y_pred_formatted = [
-                list(item[ChatCompletionConstants.OUTPUT_FULL_CONVERSATION][0].values())[0]
-                for idx, item in y_pred.iterrows()
-            ]
+            logger.info("Found more than 1 col. Trying to fetch conversation.")
+            def check_item(row_item):
```
Comment (skasturi) [inline assets/training/model_evaluation/src/evaluators/evaluators.py:621]: (nit) Add typehints & docs.

File: assets/training/model_evaluation/src/evaluators/evaluators.py [hunk 7]
```diff
@@ -617,16 +617,32 @@ def evaluate(self, y_test, y_pred, **kwargs):
...
+                item = row_item.get(ChatCompletionConstants.OUTPUT_FULL_CONVERSATION, None)
+                if not item:
```
Comment (skasturi) [inline assets/training/model_evaluation/src/evaluators/evaluators.py:623]:
```suggestion
                if item is None:
```

File: assets/training/model_evaluation/src/evaluators/evaluators.py [hunk 8]
```diff
@@ -617,16 +617,35 @@ def evaluate(self, y_test, y_pred, **kwargs):
...
+                    if item[0].get("role", False) and item[0].get("content", False):
+                        return item
+                    else:
+                        if item[0].get("0", False):
+                            return item["0"]
+                return item
+
+            y_pred_formatted = y_pred.apply(check_item, axis=1).tolist()
         # dataframe wih just predictions appended to conversations
         else:
```
Comment (skasturi) [inline assets/training/model_evaluation/src/evaluators/evaluators.py:636]: (nit) add a new line after this

File: assets/training/model_evaluation/src/evaluators/evaluators.py [hunk 9]
```diff
@@ -617,16 +617,35 @@ def evaluate(self, y_test, y_pred, **kwargs):
...
-            y_pred_formatted = [
-                list(item[ChatCompletionConstants.OUTPUT_FULL_CONVERSATION][0].values())[0]
-                for idx, item in y_pred.iterrows()
-            ]
+            logger.info("Found more than 1 col. Trying to fetch conversation.")
+
+            def check_item(row_item):
+                item = row_item.get(ChatCompletionConstants.OUTPUT_FULL_CONVERSATION, None)
+                if not item:
+                    return row_item
```
Comment (skasturi) [inline assets/training/model_evaluation/src/evaluators/evaluators.py:625]: (nit) add a new line before this

File: assets/training/model_evaluation/src/evaluators/evaluators.py (hunk 6 above)
Comment (ankushbhatia2) [reply assets/training/model_evaluation/src/evaluators/evaluators.py:622]: This "if" case is a special case we have added to handle the output from Model Prediction Component where chat completion predictions are returned in 2 formats. 1 column containing only predicted text, 2. containing full conversation.

File: assets/training/model_evaluation/src/evaluators/evaluators.py (hunk 4 above)
Comment (ankushbhatia2) [reply assets/training/model_evaluation/src/evaluators/evaluators.py:619]: Yes, y_pred is always a dataframe. We take care of that while loading the data

File: assets/training/model_evaluation/src/evaluators/evaluators.py (hunk 5 above)
Comment (ankushbhatia2) [reply assets/training/model_evaluation/src/evaluators/evaluators.py:618]: Added in latest commit

**Review by ShipraJain01 — APPROVED** [review]

**Review by skasturi — APPROVED** [review]

**Pull request merged**